### PR TITLE
Xvg 2642 python3

### PR DESCRIPTION
**Pull request by jputerova**
Please review the changes - porting dx-streaming-uploader from python2 to python3

This branch also includes changes implemented in https://github.com/dnanexus-rnd/dx-streaming-upload/tree/SF00026009_novaseq (which is waiting for approval too)

### CODE REVIEW

File: files/dx_sync_directory.py [hunk 1]
```diff
@@ -295,21 +295,21 @@ def check_inputs(args):
...
 
-    print >> sys.stderr, '\n--- Reading log file...'
+    print('\n--- Reading log file...', file=sys.stderr)
 
     if os.path.exists(args.log_file):
         with open(args.log_file, 'r') as logf:
             return json.load(logf)
     else:
-        print >> sys.stderr, '  Log file not found, returning empty log.'
+        print('  Log file not found, returning empty log.', file=sys.stderr)
```
Comment (mhrvol) [inline files/dx_sync_directory.py]:
```suggestion
        print('Log file not found, returning empty log.', file=sys.stderr)
```

File: files/dx_sync_directory.py [hunk 2]
```diff
@@ -295,21 +295,21 @@ def check_inputs(args):
...
         return {'tar_files': {}, 'next_tar_index': 0, 'files': {},
                 'tar_destination': args.tar_destination, 'file_prefix': args.prefix,
                 'sync_dir': args.sync_dir, 'include_patterns': args.include_patterns,
                 'exclude_patterns': args.exclude_patterns}
     return log
 
 def check_log(log, args):
-    print >> sys.stderr, ( '\n--- Checking that log matches inputs' )
+    print(( '\n--- Checking that log matches inputs' ), file=sys.stderr)
```
Comment (mhrvol) [inline files/dx_sync_directory.py]: Why double brackets?

File: files/dx_sync_directory.py [hunk 3]
```diff
@@ -393,9 +393,9 @@ def split_into_tar_files(files_to_upload, log, args):
     tars_to_upload.append(current_tar)
 
     if total_size < args.min_tar_size:
-        print >> sys.stderr, ('QUITTING: Size of files to upload is not big ' +
+        print(('QUITTING: Size of files to upload is not big ' +
```
Comment (mhrvol) [inline files/dx_sync_directory.py]: why double brackets? (same comment as before, just close this comment if there's an explanation)

File: files/dx_sync_directory.py [hunk 4]
```diff
@@ -460,10 +460,10 @@ def upload_tar_files(log, args):
                     opts += '--verbose '
 
                 ua_command = "ua --project %s --folder %s --do-not-compress --wait-on-close --progress %s %s --auth-token %s --chunk-size 25M" % (tar_destination_project, tar_destination_folder, opts, tar_file, args.auth_token)
-                print >> sys.stderr, ua_command
+                print(ua_command, file=sys.stderr)
                 try:
-                    dx_file_id = subprocess.check_output(ua_command, shell=True)
-                    dx_file_id = dx_file_id.strip()
+                    ua_process = subprocess.run(ua_command, shell=True, check=True, stdout=subprocess.PIPE, universal_newlines=True)
```
Comment (mhrvol) [inline files/dx_sync_directory.py:463]: Is there an advantage of using run instead of check_output?

By the way in python3 you would need to decode it, in order to get the same result as in python2.
`subprocess.check_output(ua_command, shell=True).decode('utf-8')`

File: files/incremental_upload.py [hunk 5]
```diff
@@ -467,9 +468,9 @@ def main():
         if not isinstance(input_dict, dict):
             raise_error("Expected a dict for downstream input. Got %s." %input_dict)
 
-        for k, v in input_dict.items():
-            if not ((isinstance(k, str) or isinstance(k, basestring)) and
-                    (isinstance(v, str) or isinstance(v, basestring) or isinstance(v, dict))):
+        for k, v in list(input_dict.items()):
+            if not ((isinstance(k, str) or isinstance(k, str)) and
```
Comment (mhrvol) [inline files/incremental_upload.py]: `(isinstance(k, str) or isinstance(k, str))` ... no need to use OR, right?  basestring is no longer available in python3 I believe.

File: files/monitor_runs.py [hunk 6]
```diff
@@ -134,9 +135,10 @@ def get_streaming_config(config_file, project, applet, workflow, script, token):
         config["workflow"] = workflow
     if script:
         config["script"] = os.path.abspath(script)
-    user_config_dict = yaml.load(config_file)
 
-    for key, default in CONFIG_DEFAULT.items():
+    user_config_dict = yaml.load(config_file, Loader=yaml.FullLoader)
```
Comment (mhrvol) [inline files/monitor_runs.py]: what is FullLoader needed for?

File: files/monitor_runs.py [hunk 7]
```diff
@@ -335,14 +337,15 @@ def _trigger_streaming_upload(folder, config):
...
-    command = ["python", inc_upload_script_loc,
+    command = ["python3", inc_upload_script_loc,
                "-a", config['token'],
                "-p", config['project'],
                "-r", folder,
                "-t", config['tmp_dir'],
                "-L", config['log_dir'],
                "-m", config['min_age'],
                "-z", config['min_size'],
+               "-M", config['max_size'],
```
Comment (mhrvol) [inline files/monitor_runs.py:349]: why is max_size added? Should we do this in separate commit if this is not related to python2 -> python3?

File: files/monitor_runs.py [hunk 8]
```diff
@@ -373,20 +376,20 @@ def _trigger_streaming_upload(folder, config):
...
-    print "==INFO== Triggering incremental upload command: {0}".format(
-        ' '.join(command))
+    print("==INFO== Triggering incremental upload command: {0}".format(
+        ' '.join(command)))
     try:
-        inc_out = sub.check_output(command)
-    except sub.CalledProcessError, e:
-        print "==ERROR== Incremental upload command {0} failed.\n "\
-        "Error code {1}:{2}".format(e.cmd, e.returncode, e.output)
+        inc_out = sub.run(command, check=True, stdout=sub.PIPE, universal_newlines=True).stdout
```
Comment (mhrvol) [inline files/monitor_runs.py:383]: same as above, why run instead of check_output?

File: requirements.txt [hunk 9]
```diff
@@ -0,0 +1,2 @@
+pyyaml==5.4
+dxpy
```
Comment (mhrvol) [inline requirements.txt]: should we set the version of dxpy as well?

File: tasks/main.yml [hunk 10]
```diff
@@ -22,8 +20,8 @@
   file: path=/opt/dnanexus-upload-agent state=directory mode=0755
 
 - name: Download and unzip UA tarball
-  unarchive: 
-    src: https://dnanexus-sdk.s3.amazonaws.com/dnanexus-upload-agent-1.5.31-linux.tar.gz
+  unarchive:
+    src: https://dnanexus-sdk.s3.amazonaws.com/dnanexus-upload-agent-1.5.33-linux.tar.gz
```
Comment (mhrvol) [inline tasks/main.yml]: we usually pin point versions at the top of the file, so its easier to change it. Would it make sense here?

File: tasks/main.yml [hunk 11]
```diff
@@ -50,7 +48,7 @@
 
 # Logging into DNAnexus account
 - name: Log in to DNAnexus account if dx_user_token is provided
-  shell: source /opt/dx-toolkit/environment && dx login --token {{ item.dx_user_token if 'dx_user_token' in item else dx_token }} --noprojects
+  shell: dx login --token {{ item.dx_user_token if 'dx_user_token' in item else dx_token }} --noprojects
```
Comment (mhrvol) [inline tasks/main.yml:51]: why dont we source environment anymore?

**Review by mhrvol — COMMENTED** [review]
few small comments

File: files/dx_sync_directory.py (hunk 1 above)
Comment (jputerova) [reply files/dx_sync_directory.py]: I used 2to3 library to port it to python3 and this looks like remnant of an automatic conversion

File: files/dx_sync_directory.py (hunk 2 above)
Comment (jputerova) [reply files/dx_sync_directory.py]: remnant of 2to3

File: files/dx_sync_directory.py (hunk 4 above)
Comment (jputerova) [reply files/dx_sync_directory.py:463]: the check_output returned bytes instead of string in python3 which caused several issues in the code. Instead check_output I decided to use what was recommended here for python 3.5+ https://queirozf.com/entries/python-3-subprocess-examples#run-command-and-capture-output (see section Run command and capture output)

the decode didn't work well as in other parts of the code it was complaining that it requires bytes instead of string.

File: files/monitor_runs.py (hunk 6 above)
Comment (jputerova) [reply files/monitor_runs.py]: yaml.load() without specified loader is deprecated https://github.com/yaml/pyyaml/wiki/PyYAML-yaml.load(input)-Deprecation . But looking at the page right now, they say FullLoader should be avoided for now.

File: README.md [hunk 12]
```diff
@@ -19,8 +19,11 @@ Role Variables
...
+  - `min_size`: (Optional) The minimum size of the TAR file before it will be uploaded (in MB). Default=500
+  - `max_size`: (Optional) The maximum size of the TAR file to be uploaded (in MB). Default=10000
   - `run_length`: (Optional) Expected duration of a sequencing run, corresponds to the -D paramter in incremental upload (For example, 24h). Acceptable suffix: s, m, h, d, w, M, y.
-  - `n_seq_intervals`: (Optional) Number of intervals to wait for run to complete. If the sequencing run has not completed within `n_seq_intervals` * `run_length`, it will be deemed as aborted and the program will not attempt to upload it. Corresponds to the -I parameter in incremental upload.
+  - `n_seq_intervals`: (Optional) Number of intervals to wait for run to complete. If the sequencing run has not completed within `n_seq_intervals` * `run_length`, it will be deemed as aborted and the program will not attempt to upload it. Corresponds to the -I parameter in incremental uploiad.
```
Comment (mhrvol) [inline README.md]:
```suggestion
  - `n_seq_intervals`: (Optional) Number of intervals to wait for run to complete. If the sequencing run has not completed within `n_seq_intervals` * `run_length`, it will be deemed as aborted and the program will not attempt to upload it. Corresponds to the -I parameter in incremental upload.
```

File: files/dx_sync_directory.py (hunk 2 above)
Comment (mhrvol) [reply files/dx_sync_directory.py]: I see, should we remove the double bracket?

File: files/dx_sync_directory.py (hunk 4 above)
Comment (mhrvol) [reply files/dx_sync_directory.py:463]: makes sense, thanks.

File: files/monitor_runs.py (hunk 6 above)
Comment (mhrvol) [reply files/monitor_runs.py]: I though the correct way to open yaml file is something like this:

```
with open("path") as yaml_file:
    p_dict = yaml.safe_load(yaml_file)
```

File: files/dx_sync_directory.py (hunk 3 above)
Comment (jputerova) [reply files/dx_sync_directory.py]: yes. will clean that

File: tasks/main.yml (hunk 11 above)
Comment (jputerova) [reply tasks/main.yml:51]: it is not required to load environment anymore when dx-toolkit is installed via pip3

File: files/dx_sync_directory.py (hunk 2 above)
Comment (jputerova) [reply files/dx_sync_directory.py]: yes

File: files/incremental_upload.py (hunk 5 above)
Comment (jputerova) [reply files/incremental_upload.py]: yes, no need to use the same condition twice. again, remnant of 2to3, replaced basestring by str.

File: files/monitor_runs.py (hunk 6 above)
Comment (jputerova) [reply files/monitor_runs.py]: thanks, changed

File: README.md (hunk 12 above)
Comment (jputerova) [reply README.md]: fixed

File: tasks/main.yml (hunk 10 above)
Comment (jputerova) [reply tasks/main.yml]: I will move it to vars/main.yml (as that's the way how it works for ansible)

File: requirements.txt (hunk 9 above)
Comment (jputerova) [reply requirements.txt]: version added

**Comment by jputerova** [comment]
I pushed the changes. @mhrvol please have a look.

**Review by mhrvol — APPROVED** [review]
LGTM

**Comment by mhrvol** [comment]
Was part of different PR
